### PR TITLE
Functor classes - getting started

### PR DESCRIPTION
**Pull request by DavidBauer1984**
This PR implements the basic framework for functor classes (see #3005 ) and moves the first functors, namely most of the search functors and `CalcDots` as well as `CalcLedgerLines`.

- We have const functors (class `ConstFunctor`) and mutable functors (class `MutableFunctor`). The latter are allowed to make changes on elements that they visit. Both are derived from `FunctorBase` which stores basic properties such as traversal direction, the current functor code or filters. They are derived from different interfaces (`ConstFunctorInterface` and `FunctorInterface`) which contain the virtual methods for element visitation. For convenience, we have the classes `DocFunctor` and `DocConstFunctor` which additionally store a Doc pointer.
- For now we implement visitation for `Object`, `Doc`, all layer elements and all container elements. This will be further extended in the future.
- The search functors demonstrate the implementation and usage of simple functor classes.
- `CalcDots` demonstrates the implementation of an advanced functor with private helper function.
- `CalcLedgerLines` demonstrates the implementation of an advanced functor that also uses the end interface (formerly end functor). The call of functors now becomes quite neat: 
```cpp
CalcLedgerLines calcLedgerLines(doc);
this->Process(calcLedgerLines);
```
- Changing the code structure already becomes apparent in `note.cpp` which is reduced by ~100 LOC.

This PR should not trigger any changes in performance or rendering.

### CODE REVIEW

**Comment by lpugin** [comment]
Thanks! I changed based to `develop-functors`.

I am planning to have a release with 3.12 this week. We can start merging this after that.

**Comment by DavidBauer1984** [comment]
@lpugin Release build size (from Xcode) is now 17.855 MB compared to 17.620 MB before.

File: src/object.cpp [hunk 1]
```diff
@@ -1104,6 +1104,140 @@ void Object::Process(Functor *functor, FunctorParams *functorParams, Functor *en
...
+                if (this->FiltersApply(filters, *iter)) {
+                    (*iter)->Process(functor, deepness);
+                }
+            }
+        }
+    }
+
+    if (functor.ImplementsEndInterface() && !skipFirst) {
+        FunctorCode code = this->VisitEnd(functor);
+        functor.SetCode(code);
```
Comment (lpugin) [inline src/object.cpp:1160]: Why not setting the `code` directly in `Visit` or `VisitEnd` methods? Then you would do just
```cpp
this->VisitEnd(functor);
```

Comment (DavidBauer1984) [reply src/object.cpp]: We would have to set the code in each override of `Visit(End)`, now `Accept(End)`. So we would trade setting it twice in `Process` against setting it 250 times in `Accept`.

**Comment by DavidBauer1984** [comment]
@lpugin Is there anything left to do here?

**Pull request merged**